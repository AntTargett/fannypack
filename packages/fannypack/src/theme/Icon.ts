// @ts-ignore
import _get from 'lodash/get';
import * as faInfoCircle from '@fortawesome/free-solid-svg-icons/faInfoCircle';
import * as faExclamationTriangle from '@fortawesome/free-solid-svg-icons/faExclamationTriangle';
import * as faCheckCircle from '@fortawesome/free-solid-svg-icons/faCheckCircle';
import * as faExclamationCircle from '@fortawesome/free-solid-svg-icons/faExclamationCircle';
import * as faQuestionCircle from '@fortawesome/free-solid-svg-icons/faQuestionCircle';
import * as faTimes from '@fortawesome/free-solid-svg-icons/faTimes';
import * as faSearch from '@fortawesome/free-solid-svg-icons/faSearch';

import { parseIcons, Opts as ParseIconsOpts, Icons } from '../utils/parseIcons';

const parseOverrideIcons = (
  icons: Array<{ icons: Icons; type: ParseIconsOpts['type']; prefix: ParseIconsOpts['prefix'] }>
) =>
  icons.reduce(
    (currentIcons: {}, iconSet: { icons: Icons; type: ParseIconsOpts['type']; prefix: ParseIconsOpts['prefix'] }) => ({
      ...currentIcons,
      ...parseIcons(iconSet.icons, { type: iconSet.type, prefix: iconSet.prefix })
    }),
    {}
  );

export default (overrides: any) => ({
  ...overrides,
  icons: {
    ...parseIcons(
      [faInfoCircle, faExclamationTriangle, faCheckCircle, faExclamationCircle, faQuestionCircle, faTimes, faSearch],
      {
        type: 'font-awesome-standalone'
      }
    ),
    ...parseOverrideIcons(_get(overrides, 'iconSets', [])),
    ..._get(overrides, 'icons', {})
  },
  iconNames: {
    info: 'info-circle',
    warning: 'exclamation-triangle',
    success: 'check-circle',
    danger: 'exclamation-circle',
    ..._get(overrides, 'iconNames', {})
  }
});
