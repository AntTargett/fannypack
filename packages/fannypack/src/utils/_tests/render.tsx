import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
