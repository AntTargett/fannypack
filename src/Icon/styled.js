import { Box } from '../primitives';
import styled from '../styled';
import { palette, theme } from 'styled-tools';

export default styled(Box)`
  fill: ${props => palette(props.color || 'text')(props)};
  position: relative;
  top: 0.1em;
  ${theme('fannypack.Icon.base')};
`;
