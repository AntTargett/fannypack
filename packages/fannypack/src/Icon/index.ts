import Icon from './Icon';
import * as styledIcon from './styled';

export default Icon;

export * from './Icon';
export { default as Icon } from './Icon';

export { styledIcon };
