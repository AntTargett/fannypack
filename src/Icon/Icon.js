// @flow
import React, { type Node } from 'react';
import { withTheme } from 'reakit/styled';
import * as icons from '@fortawesome/free-solid-svg-icons';
import _get from 'lodash/get';
import _upperFirst from 'lodash/upperFirst';
import _camelCase from 'lodash/camelCase';

import _Icon from './styled';
import type { Size } from '../types';

type Props = {
  /** A label for the icon which can be read by screen readers. This is required! */
  a11yLabel: string,
  children: Node,
  /** Color of the icon. Can be a color from the palette, or any other color. */
  color: string,
  className?: string,
  /** The name of your icon from the free Font Awesome Icon Set (https://fontawesome.com/icons?d=gallery&m=free). */
  icon: string,
  /** Size of the icon. Available values: "small", "medium", "large" */
  size?: Size,
  theme: Object
};

export const Icon = ({ a11yLabel, children, icon, size: _size, theme, ...props }: Props) => {
  const size = _get(theme, `fannypack.fontSizes[${_size || ''}]`, 1);
  const newIcon = _get(theme, `fannypack.Icon.iconNames[${icon}]`) || icon;
  const iconInfo = icons[`fa${_upperFirst(_camelCase(newIcon))}`];
  const [viewBoxWidth, viewBoxHeight, , , iconPath] = iconInfo.icon;
  return (
    <_Icon use="svg" role="img" size={size} viewBox={`0 0 ${viewBoxWidth} ${viewBoxHeight}`} {...props}>
      {a11yLabel && <title>{a11yLabel}</title>}
      <path d={iconPath} fill="currentColor" />
    </_Icon>
  );
};

Icon.defaultProps = {
  a11yLabel: undefined,
  className: undefined,
  color: undefined,
  icon: undefined,
  size: 'default'
};

export default withTheme(Icon);
