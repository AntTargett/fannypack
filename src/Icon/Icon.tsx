import * as React from 'react';
import * as PropTypes from 'prop-types';
// @ts-ignore
import propTypeUtils from 'airbnb-prop-types';
import { BoxProps as ReakitBoxProps } from 'reakit/ts';
import * as icons from '@fortawesome/free-solid-svg-icons';
// @ts-ignore
import _get from 'lodash/get';
// @ts-ignore
import _upperFirst from 'lodash/upperFirst';
// @ts-ignore
import _camelCase from 'lodash/camelCase';

import { withTheme } from '../styled';
import { Omit, Size, sizePropType } from '../types';
import _Icon from './styled';

export type Props = {
  children?: React.ReactNode;
  /** Color of the icon. Can be a color from the palette, or any other color. */
  color?: string;
  className?: string;
  /** The name of your icon from the free Font Awesome Icon Set (https://fontawesome.com/icons?d=gallery&m=free). */
  icon: string;
  /** Size of the icon. Available values: "small", "medium", "large" */
  size?: Size;
  theme?: Object;
};
export type PropsWithA11yHidden = Props & {
  /** Indicates that this element should be skipped by assistive technologies. */
  a11yHidden: boolean;
};
export type PropsWithA11yLabel = Props & {
  /** A label for the icon which can be read by screen readers. This is required if a11yHidden is false. */
  a11yLabel: string;
};
export type LocalIconProps = PropsWithA11yHidden | PropsWithA11yLabel;
export type IconProps = Omit<ReakitBoxProps, 'size'> & LocalIconProps;

export const Icon: React.FunctionComponent<LocalIconProps> = ({
  // @ts-ignore
  a11yHidden,
  // @ts-ignore
  a11yLabel,
  children,
  icon,
  size: _size,
  theme,
  ...props
}) => {
  const size = _get(theme, `fannypack.fontSizes[${_size || ''}]`, 1);
  const newIcon = _get(theme, `fannypack.Icon.iconNames[${icon}]`) || icon;
  // @ts-ignore
  const iconInfo = icons[`fa${_upperFirst(_camelCase(newIcon))}`];
  const [viewBoxWidth, viewBoxHeight, , , iconPath] = iconInfo.icon;
  return (
    // @ts-ignore
    <_Icon
      use="svg"
      ariaHidden={a11yHidden}
      role="img"
      size={size}
      viewBox={`0 0 ${viewBoxWidth} ${viewBoxHeight}`}
      {...props}
    >
      {a11yLabel && <title>{a11yLabel}</title>}
      <path d={iconPath} fill="currentColor" />
    </_Icon>
  );
};

Icon.propTypes = {
  a11yHidden: propTypeUtils.mutuallyExclusiveProps(PropTypes.bool, 'a11yHidden', 'a11yLabel'), // eslint-disable-line
  a11yLabel: propTypeUtils.mutuallyExclusiveProps(PropTypes.string, 'a11yHidden', 'a11yLabel'), // eslint-disable-line
  children: PropTypes.node,
  color: PropTypes.string,
  className: PropTypes.string,
  icon: PropTypes.string.isRequired,
  size: sizePropType,
  theme: PropTypes.object // eslint-disable-line
};
Icon.defaultProps = {
  children: null,
  className: undefined,
  color: undefined,
  icon: undefined,
  size: 'default'
};

// @ts-ignore
const C: React.FunctionComponent<IconProps> = withTheme(Icon);
export default C;
