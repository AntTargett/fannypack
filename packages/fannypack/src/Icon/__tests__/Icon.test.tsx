import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Icon } from '../Icon';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Icon>Hello world</Icon>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Icon color="primary">Hello world</Icon>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(<Icon use="div">Hello world</Icon>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Icon props', () => {
      const { result } = renderHook(() => Icon.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(<Icon>{iconProps => <div {...iconProps}>Hello world</div>}</Icon>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('Icon.base should render correctly', () => {
    const { container } = render(<Icon>hello world</Icon>, {
      theme: { Icon: { base: { backgroundColor: 'red' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<Icon>hello world</Icon>, {
      theme: { Icon: { defaultProps: { className: 'test' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
