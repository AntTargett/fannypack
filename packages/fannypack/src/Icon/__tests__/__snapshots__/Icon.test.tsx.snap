// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  fill: inherit;
  position: relative;
  height: 1em;
  width: 1em;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Icon"
  role="img"
  viewbox="0 0 0 0"
/>
`;

exports[`composition hook should return with Icon props 1`] = `
Object {
  "children": <React.Fragment />,
  "className": "css-i5runh css-1pwrqjj css-1ccgglo fp-Box fp-Icon",
  "role": "img",
  "viewBox": "0 0 0 0",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  fill: inherit;
  position: relative;
  height: 1em;
  width: 1em;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Icon"
  role="img"
  viewbox="0 0 0 0"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  fill: inherit;
  position: relative;
  height: 1em;
  width: 1em;
}

<svg
  class="emotion-0 emotion-1 emotion-2 test fp-Box fp-Icon"
  role="img"
  viewBox="0 0 0 0"
/>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  fill: inherit;
  position: relative;
  height: 1em;
  width: 1em;
}

<svg
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Icon"
  role="img"
  viewBox="0 0 0 0"
/>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2.emotion-2.emotion-2 {
  color: #3722ed;
}

.emotion-0 {
  fill: #3722ed;
  position: relative;
  height: 1em;
  width: 1em;
}

<svg
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Icon"
  role="img"
  viewBox="0 0 0 0"
/>
`;

exports[`theming Icon.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  fill: inherit;
  position: relative;
  height: 1em;
  width: 1em;
}

.emotion-0 {
  background-color: red;
}

<svg
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Icon"
  role="img"
  viewBox="0 0 0 0"
/>
`;
